hardware/detect.py

```diff
@@ -710,85 +731,88 @@ def get_cpus(hw_lst):
+                                     ('flags', 'Flags', None)]:
+            value = None
+            if d_key in lscpu:
+                value = lscpu[d_key]
+                if conv:
+                    value = conv(value)
+            elif t_key == 'threads':
+                value = (int(lscpu.get('Thread(s) per core', 1)) *
+                         int(lscpu.get('Core(s) per socket', 1)))
+            if value is not None:
+                hw_lst.append(('cpu', ptag, t_key, value))
 
     hw_lst.append(('cpu', 'logical', 'number', int(lscpu['CPU(s)'])))
-    # Govenors could be differents on logical cpus
+    # Governors could be different on logical cpus
     for cpu in range(int(lscpu['CPU(s)'])):
-        scaling_governor = "/sys/devices/system/cpu/cpufreq/policy{}/scaling_governor".format(
-            cpu)
-        if os.path.exists(scaling_governor):
-            with open(scaling_governor) as governorfile:
-                hw_lst.append(('cpu', "logical_{}".format(cpu),
-                               "governor", governorfile.readline().rstrip('\n')))
+        ltag = "logical_{}".format(cpu)
+        (exists, value) = _from_file(("/sys/devices/system/cpu/cpufreq/"
+                                      "policy{}/scaling_governor".format(cpu)))
+        if exists:
+            hw_lst.append(('cpu', ltag, "governor", value))
 
     # Extracting numa nodes
     hw_lst.append(('numa', 'nodes', 'count', int(lscpu['NUMA node(s)'])))
-    for numa in range(int(lscpu['NUMA node(s)'])):
-        cpus = lscpu['NUMA node{} CPU(s)'.format(numa)]
+
+    # Allow for sparse numa nodes
+    # NOTE(tonyb): When extracting the numeric id we remove the 'NUMA node' and
+    # ' CPUS(s)' so that 'NUMA node1024 CPU(s)' becomes '1024 '
+    # NOTE(tonyb): Use sorted() to ensure we get a stable order, just in case
+    # consumers of hw_lst rely on that
+    numa_nodes = [(k, int(k[9:-7]))
+                  for k in lscpux.keys()
+                  if 'NUMA node' in k and 'CPU(s)' in k]
+    numa_nodes = sorted(numa_nodes, key=lambda t: int(t[1]))
+    for (key, node_id) in numa_nodes:
+        ntag = 'node_{}'.format(node_id)
+        cpus = lscpu[key]
         # lscpu -x provides the cpu mask
-        cpu_mask = lscpux['NUMA node{} CPU(s)'.format(numa)]
+        cpu_mask = lscpux[key]
         total_cpus = 0
         min_cpu = None
         max_cpu = None
 
-        for item in cpus.split(','):
-            # lscpu report numa nodes like 0-5,48-53
-            if "-" in item:
-                max_cpu = int(item.split("-")[1])
-                min_cpu = int(item.split("-")[0])
-                total_cpus = total_cpus + max_cpu - min_cpu + 1
-            else:
-                # or like 0,1
-                if min_cpu is None:
-                    min_cpu = int(item)
-                else:
-                    max_cpu = int(item)
+        # It's possible to have a NUMA node without any CPUs
+        if cpus:
```

Comment:
nit:
```
if not cpus:
    continue
```

Reply:
That wont work as we'll fail to add the empty numa node into the list.